include/tvm/runtime/container.h

```diff
@@ -188,6 +188,709 @@ class InplaceArrayBase {
+   * \param other The source of assignment
+   * \return reference to self.
+   */
+  Array<T>& operator=(const Array<T>& other) {
+    data_ = other.data_;
+    return *this;
+  }
+
+ public:
+  // iterators
+  struct ValueConverter {
+    using ResultType = T;
+    static inline T convert(const ObjectRef& n) { return DowncastNoCheck<T>(n); }
+  };
+
+  using iterator = IterAdapter<ValueConverter, const ObjectRef*>;
+  using reverse_iterator = ReverseIterAdapter<ValueConverter, const ObjectRef*>;
+
+  /*! \return begin iterator */
+  inline iterator begin() const { return iterator(GetArrayNode()->begin()); }
+
+  /*! \return end iterator */
+  inline iterator end() const { return iterator(GetArrayNode()->end()); }
+
+  /*! \return rbegin iterator */
+  inline reverse_iterator rbegin() const { return reverse_iterator(GetArrayNode()->end() - 1); }
+
+  /*! \return rend iterator */
+  inline reverse_iterator rend() const { return reverse_iterator(GetArrayNode()->begin() - 1); }
+
+ public:
+  // const methods in std::vector
+  /*!
+   * \brief Immutably read i-th element from array.
+   * \param i The index
+   * \return the i-th element.
+   */
+  inline const T operator[](const int64_t i) const {
+    ArrayNode* data = GetArrayNode();
+    CHECK(data != nullptr) << "ValueError: cannot index a null Array";
+    return DowncastNoCheck<T>((*data)[i]);
+  }
+
+  /*! \return The size of the array */
+  inline size_t size() const {
+    ArrayNode* p = GetArrayNode();
+    return p == nullptr ? 0 : GetArrayNode()->size_;
+  }
+
+  /*! \return The capacity of the array */
+  inline size_t capacity() const {
+    ArrayNode* p = GetArrayNode();
+    return p == nullptr ? 0 : GetArrayNode()->capacity_;
+  }
+
+  /*! \return Whether array is empty */
+  inline bool empty() const { return size() == 0; }
+
+ public:
+  // mutation in std::vector, implements copy-on-write
+
+  /*! \brief Release reference to all the elements */
+  inline void clear() {
+    if (data_ != nullptr) {
+      ArrayNode* p = CopyOnWrite();
+      p->clear();
+    }
+  }
+
+  /*!
+   * \brief Make sure the list has the capacity of at least n
+   * \param n lower bound of the capacity
+   */
+  inline void reserve(const int64_t n) {
+    if (data_ == nullptr || n > GetArrayNode()->capacity_) {
+      SwitchContainer(n);
+    }
+    CHECK_GE(GetArrayNode()->capacity_, n);
+  }
+
+  /*!
+   * \brief Resize the array.
+   * \param n The new size.
+   */
+  inline void resize(const int64_t n) {
+    // corner case: do nothing
+    if (data_ == nullptr) {
+      SwitchContainer(n);
+    }
+    if (GetArrayNode()->size_ == n) {
+      return;
+    }
+    // ensure enough capacity
+    CHECK_GE(n, 0) << "ValueError: cannot resize an Array to negative size";
+    reserve(n);
+    // mutate by expanding or shrinking
+    ArrayNode* p = CopyOnWrite();
+    ObjectRef* iter = p->MutableBegin() + (p->size_ - 1);
+    int64_t& cur_size = p->size_;
+    if (n > cur_size) {
```

Comment:
Consider to make it as a member function of ArrayNode->resize, allow pass in the default value in.

Reply:
Alternative, we can have two calls, Enlarge(bigger_size, default_value), Shrink(target_size)